src/abstractdataframe/iteration.jl

```diff
@@ -454,6 +454,16 @@ function mapcols!(f::Union{Function, Type}, df::DataFrame)
     if len_min != len_max
         throw(DimensionMismatch("lengths of returned vectors must be identical"))
     end
+
+    for (i, col) in enumerate(vs)
+        fic = firstindex(col)
+        if fic != 1
+            throw(ArgumentError("Currently DataFrames.jl supports only " *
+                                "columns that use 1-based indexing and " *
+                                "column $i has starting index equal to $fic"))
+        end
```

Comment:
How about moving this to a separate function to avoid repeating the error message?

Reply:
OK - makes sense. Actually I think we should even remove all uses of the `_columns` function as using it leads to unsafe code. I will work on this. I understand you agree with a general direction I propose here.

Reply:
I will leave `_columns` for now as we have different use-cases for it. I have added a developers comment to it though